Standardize times in time sensors Jewish calendar

tsvi opened the pull request:
## Breaking Change:

The output of the timestamp sensors have been streamlined so they're easier to use in automations. All the timestamp sensors will return UTC time in ISO 8601 format.
Attributes have been added to get a UNIX timestamp.

**Please note if you have any automations relying on the timestamp, you'll need to update them to compare to the UTC timestamp.**

## Description:
This standardizes the times in the timestamp sensors for the Jewish calendar, times like candle lighting, havdalah etc. are all defined as timestamp sensors. As such, their state will show ISO8601 UTC timestamps.
Also, timestamp attributes have been added to the sensor. This should allow easier automation triggers and conditions when comparing timestamps.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10463

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

tsvi committed:
Standardize times in time sensors Jewish calendar

tsvi committed:
Fix pylint errors

tsvi committed:
Add non-default time format test

tsvi committed:
Make black happy

MartinHjelmare commented on homeassistant/components/jewish_calendar/sensor.py:
```diff
@@ -92,45 +83,101 @@ def make_zmanim(date):
+    def state(self):
+        """Return the state of the sensor."""
+        return (
+            self._state.strftime(self._time_format)
```
Sensors with timestamp as device class must report an isoformatted absolute utc timestamp as state. That doesn't seem the case here, right?

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

tsvi replied:
ISO 8601 doesn't relate to UTC. But, you're right the return value depends on the configuration.

I'll remove the device_class

tsvi replied:
@MartinHjelmare pushed a fix

tsvi replied:
@MartinHjelmare, I'm actually not completely convinced this shouldn't be a timestamp device class: By default I show %H:%M, but the datetime object which defined the state returns all of the data necessary to show it as a full timestamp.

Maybe the key is wrong as it defines a representation, so that should probably be part of Lovelace. And then I return a full timestamp (localized but ISO8601 compliant), but then I'd like a way to hint to Lovelace that the default representation should be %H:%M.

By the way I think the test errored due to something flaky.

Thanks,

Tsvi

MartinHjelmare replied:
We should keep the timestamp device class and make sure the state is isoformatted utc. The frontend representation is a frontend matter.

Currently only the user have some options to change the frontend formatting:
https://www.home-assistant.io/lovelace/entities/#format

balloob replied:
Home Assistant can represent a timestamp as just HH:MM: https://www.home-assistant.io/lovelace/entities/#format

tsvi replied:
@balloob it looks like there is no option for HH:MM. If the use chooses 'time' I suppose it will render with the seconds as well?

Also, is there a way the entity can hint to Lovelace what the default should be? Obviously if the user wants to represent the sensor differently that's their issue.

My reasoning is that I don't want the user to rely on the seconds precision, so I don't want to show it by default.

balloob replied:
There is currently no way to provide a hint to Lovelace. 

The advantage of using the timestamp device class is that users could have a countdown to the time. I am however not familiar enough to know if that is relevant in this case. Or would everyone always just want to see HH:MM ?

tsvi replied:
It definitely is relevant. Could be a nice touch for a countdown towards candle lighting especially.

I reverted the removal of the timestamp device class. The state is a datetime which is timezone aware, so it should show up correctly (or should I call `self._state.isoformat() if not None` and removed the time format configuration, as it should be solved in the frontend.

I'm wondering if we should add the attributes I added here to all device_class == "timestamp"? Could have a nice use case for working with input_datetime/sensor.time?

balloob replied:
The timestamp should be UTC, the browser will convert it to the relevant local time.

tsvi replied:
Won't that cause issues with automations trying to compare to `sensor.time`? Or how else would I check if a given time has arrived?

balloob replied:
`utcnow()` is available in templates in automations.

tsvi replied:
@balloob, @MartinHjelmare I checked the UI, it doesn't change if the state is in UTC or not. As long as it's not a datetime naive result the UI will show it correctly.
It would be better to leave it as non-UTC as it makes more sense when looking at the data. (When clicking on the sensor it looks weird seeing the time in UTC while the time in the general UI is local, you could compare it to looking at a timestamp for the next bus arrival).
So unless I hear a strong case for changing it to UTC, I'm inclined to leave it localized.

ISO8601 does *not* specify UTC timezone, rather it specifies the format *how* the timezone data is to be represented.

MartinHjelmare replied:
All our times in the state and state attributes should be utc. That makes it easy to reason about and users know what to expect.

The frontend is a separate concern.

tsvi replied:
Ok, so in the case of a timestamp showing the bus arrival, the state should also UTC time?

MartinHjelmare replied:
Yes.

tsvi committed:
Remove timestamp device class

Timestamp device class requires ISO 8601 format

mikeage commented:
Nice idea. We make "early shabbat" during DST (well, mostly; we stopped a few weeks ago), so I created my own issur_melacha sensor based on plag hamincha on Friday, and the conversion to compare the current time with plag as a timestamp was a bit of an exercise in Jinja and figuring out what's what. This will make that sort of thing much easier :-)

tsvi committed:
Revert "Remove timestamp device class"

This reverts commit 8a2fda39831bc750c3a77aa774b84b054d78032c.

tsvi committed:
Remove time_format

As this is part of the UI decision, it should be decided by lovelace.

A nice addition for a future PR, might be the option to hint to lovelace the preferred way to display some data.

MartinHjelmare commented on homeassistant/components/jewish_calendar/sensor.py:
```diff
@@ -92,45 +82,97 @@ def make_zmanim(date):
+        return "timestamp"
+
+    @property
+    def state_attributes(self):
```
Please rename to `device_state_attributes`.

tsvi committed:
Update name of state_attributes

tsvi commented:
I'm closing this pull request for now. I'm trying to envision what I would like ultimately to be possible and I think it will need some other PR's first.

Ultimately, I would like to reach a point where I could write the following automation:
```
automation:
  trigger:
    platform: time
    at: "{{ states('sensor.timestamp_based_sensor') }}"
    offset:
      minutes: -15
```

@balloob would you consider such a PR? Or am I in the wrong direction, I suppose this would mean opening up an architecture issue?

balloob commented:
Yeah I would consider such a PR . It might be better to just have time trigger be aware of timestamp device classes sensors.  That way the UI could offer a dropdown.

tsvi committed:
State of timestamp variable to be shown in UTC

Although I don't understand it, I give up :)

MartinHjelmare commented on homeassistant/components/jewish_calendar/sensor.py:
```diff
@@ -92,45 +82,97 @@ def make_zmanim(date):
+        attrs["year"] = self._state.year
+        attrs["month"] = self._state.month
+        attrs["day"] = self._state.day
+        attrs["hour"] = self._state.hour
```
Do these times also need changing to reflect utc time?

tsvi replied:
I removed them, don't see the value where one would just want the hour of one of these times.
Also, I think, if implemented, these should go in the sensor component for the timestamp device class.

tsvi committed:
Remove unnecessary attributes

I don't really see the value in these attributes, if there are any they should be implemented in
the sensor component for the timestamp device class

MartinHjelmare commented:
Thanks!
